### PR TITLE
feat: add data layer and model for administrative instance messages

### PR DESCRIPTION
**Pull request by buddy326**

### CODE REVIEW

File: virtool/messages/oas.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+from pydantic import BaseModel, validator
+
+
+class CreateMessageRequest(BaseModel):
+    color: str
+    message: str
+
+
+class UpdateMessageRequest(BaseModel):
+    color: Optional[str]
```
Comment (igboyes) [inline virtool/messages/oas.py:12]: This needs to be `MessageColor`.

File: virtool/messages/models.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+from sqlalchemy import Column, Integer, Boolean, Enum, String, DateTime
+
+from virtool.pg.base import Base
+from virtool.pg.utils import SQLEnum
+
+
+class Color(str, SQLEnum):
```
Comment (igboyes) [inline virtool/messages/models.py:7]: Can we reuse core `MessageColor` here?

File: virtool/messages/data.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+            if data.color is not None:
+                instance_message.color = data.color
+
+            if data.message is not None:
+                instance_message.message = data.message
+                instance_message.updated_at = virtool.utils.timestamp()
+
+            if data.active is not None:
+                instance_message.active = False
```
Comment (igboyes) [inline virtool/messages/data.py:84]: Is this really consistent with how you have previously written request handlers with optional values?

* https://github.com/virtool/virtool-core/blob/bf9097278b08990f0c72fa99f1852b8cc6d7e562/virtool_core/models/validators.py#L36
* https://github.com/virtool/virtool/pull/2718/commits/faf3731779c0c872b7943312d3c4a8f58ddbb48b

File: virtool/messages/data.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        Update the active administrative instance message.
+        """
+
+        async with AsyncSession(self._pg) as session:
+            instance_message = (
+                await session.execute(select(SQLInstanceMessage).order_by(SQLInstanceMessage.id.desc()))
+            ).first()[0]
+
+            if not instance_message:
+                raise ResourceNotFoundError()
```
Comment (igboyes) [inline virtool/messages/data.py:71]:
```suggestion
                raise ResourceNotFoundError
```

File: virtool/messages/data.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        return InstanceMessage(**document)
+
+    async def update(self, data: UpdateMessageRequest) -> InstanceMessage:
+        """
+        Update the active administrative instance message.
+        """
+
+        async with AsyncSession(self._pg) as session:
+            instance_message = (
+                await session.execute(select(SQLInstanceMessage).order_by(SQLInstanceMessage.id.desc()))
```
Comment (igboyes) [inline virtool/messages/data.py:67]: Good work on this!

File: virtool/messages/oas.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    message: Optional[str]
+    active: Optional[bool]
+
+    @validator("active")
+    def active_must_be_false(cls, active: bool) -> bool:
+        if active:
+            raise ValueError("active can only be `False` when updating")
+        return active
```
Comment (igboyes) [inline virtool/messages/oas.py:20]: Please double-check this can't be done with `Field`.

**Review by igboyes — CHANGES_REQUESTED** [review]
There is a slip-up on `PATCH` request handling, but this is one of your best PRs so far 👍. Really clean and tests are intuitive.

File: virtool/messages/oas.py (hunk 6 above)
Comment (buddy326) [reply virtool/messages/oas.py:20]: it cant, theres not a lot of options for boolean unfortunately.

File: virtool/messages/data.py (hunk 3 above)
Comment (buddy326) [reply virtool/messages/data.py:84]: little confused by this comment, could you expand?

Comment (igboyes) [reply virtool/messages/data.py:84]: Use `exclude_unset=True` and `prevent_none`.

**Review by igboyes — APPROVED** [review]

**Pull request merged**